ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+        try:
+            response = await handler(request)
+        except Exception as err:
+            self.logger.error("Error handling request: %s  error: %s", request, err)
+            self.logger.debug("Error handling request: %s", request, exc_info=True)
+            response = generate_response(request, None, f"Unexpected Error: {err}")
+        finally:
+            return json.dumps(response)
+
+    async def serve(self, ipc_path: pathlib.Path) -> None:
+        self.logger.info("Starting RPC server over IPC socket: %s", ipc_path)
+
+        with trio.socket.socket(trio.socket.AF_UNIX, trio.socket.SOCK_STREAM) as sock:
+            # TODO: unclear if the following stuff is necessary:
+            # ###################################################
+            # These options help fix an issue with the socket reporting itself
+            # already being used since it accepts many client connection.
+            # https://stackoverflow.com/questions/6380057/python-binding-socket-address-already-in-use
+            # sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            # ###################################################
+            await sock.bind(str(ipc_path))
+            # Allow up to 10 pending connections.
+            sock.listen(10)
+
+            self._serving.set()
+
+            while self.manager.is_running:
+                conn, addr = await sock.accept()
+                self.logger.debug("Server accepted connection: %r", addr)
+                self.manager.run_task(self._handle_connection, conn)
+
+    async def _handle_connection(self, socket: trio.socket.SocketType) -> None:
+        buffer = io.StringIO()
+
+        with socket:
+            while True:
+                request = await read_json(socket, buffer)
+
+                if not isinstance(request, collections.abc.Mapping):
+                    logger.debug("Invalid payload: %s", type(request))
+                    await write_error(socket, "Invalid Request: not a mapping")
+                    continue
+
+                if not request:
+                    self.logger.debug("Client sent empty request")
+                    await write_error(socket, "Invalid Request: empty")
+                    continue
+
+                try:
+                    validate_request(request)
+                except ValidationError as err:
+                    await write_error(socket, str(err))
+                    continue
+
+                try:
+                    result = await self.execute_rpc(cast(RPCRequest, request))
+                except Exception as e:
+                    self.logger.exception("Unrecognized exception while executing RPC")
+                    await write_error(socket, "unknown failure: " + str(e))
+                else:
+                    if not result.endswith(NEW_LINE):
+                        result += NEW_LINE
+
+                    try:
+                        await socket.send(result.encode())
```

Comment:
I don't know what the correct thing to do is, but I notice this is a naked `encode()`, while [above](https://github.com/ethereum/ddht/pull/90/files#diff-b3928bcb6f44d42e17823787bdeabc5aR74) the code explicitly uses utf-8. Are these intentionally inconsistent?